File: Java/src/main/java/com/DSA/Java/Queue/CircularArrayQueue.java
package com.DSA.Java.Queue;

public class CircularArrayQueue {
   public int[] arr;
   public int startOfQueue;
   public int endOfQueue;

   public CircularArrayQueue(int size) {
      this.arr = new int[size];
      startOfQueue = endOfQueue = -1;
   }

   public boolean isEmpty() {
      if (endOfQueue == -1) {
         return true;
      }
      return false;
   }

   public boolean isFull() {
      if (endOfQueue + 1 == startOfQueue) {
         return true;
      } else if (startOfQueue == 0 && endOfQueue == arr.length - 1) {
         return true;
      }
      return false;
   }

   public void enQueue(int value) {
      if (isEmpty()) {
         arr[0] = value;
         startOfQueue = endOfQueue = 0;
      } else {
         if (isFull()) {
            System.out.println("The Queue is full");
            return ;
         }  
         if (endOfQueue + 1 == arr.length) {
            endOfQueue = 0;
         } else {
            endOfQueue++;
         }
         arr[endOfQueue] = value;
      }
   }

   public int deQueue() {
      if (isEmpty()) {
         System.out.println("The Queue is already empty");
         return -1;
      }
      int result = arr[startOfQueue];
      startOfQueue++;
      if (startOfQueue > endOfQueue) {
         startOfQueue = endOfQueue = -1;
      } else if (startOfQueue == arr.length - 1) {
         startOfQueue = 0;
      }
      return result;
   }

   public int peek() {
      if (isEmpty()) {
         System.out.println("The Queue is already empty");
         return -1;
      }
      return arr[startOfQueue];
   }

   public void delete() {
      arr = null;
      System.out.println("The Queue was deleted successfully");
   }
}

File: C++/Queue/src/CircularArrayQueue.cpp
#include "../include/CircularArrayQueue.h"

CircularArrayQueue::CircularArrayQueue(int size) {
    this->size = size;
    this->queue = new int[size];
    this->start_of_queue = this->end_of_queue = -1;
}

bool CircularArrayQueue::is_full() const {
    if (end_of_queue + 1 == start_of_queue) {
        return true;
    } else if (start_of_queue == 0 && end_of_queue == size - 1) {
        return true;
    }
    return false;
}

bool CircularArrayQueue::is_empty() const {
    return end_of_queue < 0;
}

void CircularArrayQueue::en_queue(int value) {
    if (is_empty()) {
        queue[0] = value;
        start_of_queue = end_of_queue = 0;
    } else {
        if (is_full()) {
            std::cout << "The Queue is full" << std::endl;
        } else {
            if (end_of_queue < size - 1) {
                end_of_queue++;
            } else {
                end_of_queue = 0;
            }
            queue[end_of_queue] = value;
        }
    }
}

int CircularArrayQueue::de_queue() {
    if (is_empty()) {
        std::cout << "The Queue is already empty" << std::endl;
        return -1;
    } 
    int result = queue[start_of_queue];
    start_of_queue++;
    if (start_of_queue > end_of_queue) {
        start_of_queue = end_of_queue = -1;
    } else if (start_of_queue == size - 1) {
        start_of_queue = 0;
    }
    return result;
}

int CircularArrayQueue::peek() const {
    if (is_empty()) {
        std::cout << "The Queue is empty" << std::endl;
        return -1;
    } 
    return queue[start_of_queue];
}

void CircularArrayQueue::delete_queue() {
    start_of_queue = end_of_queue = -1;
}

CircularArrayQueue::~CircularArrayQueue() {
    delete [] queue;
}
